File: src/main/java/ordilov/lissn/like/domain/LikeServiceImpl.java
package ordilov.lissn.like.domain;

import java.util.Optional;
import lombok.RequiredArgsConstructor;
import ordilov.lissn.like.domain.LikeCommand.LikePlaylistRequest;
import ordilov.lissn.like.domain.LikeCommand.LikeTrackRequest;
import ordilov.lissn.like.domain.LikeInfo.LikedPlaylistInfo;
import ordilov.lissn.like.domain.LikeInfo.LikedTrackInfo;
import ordilov.lissn.member.domain.Member;
import ordilov.lissn.member.application.port.out.MemberReader;
import ordilov.lissn.playlist.domain.Playlist;
import ordilov.lissn.playlist.domain.PlaylistReader;
import ordilov.lissn.track.domain.Track;
import ordilov.lissn.track.domain.TrackReader;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;


@Service
@Transactional
@RequiredArgsConstructor
public class LikeServiceImpl implements LikeService {

  private final LikeStore likeStore;
  private final LikeReader likeReader;
  private final TrackReader trackReader;
  private final MemberReader memberReader;
  private final PlaylistReader playlistReader;

  @Override
  public LikedPlaylistInfo like(LikePlaylistRequest request) {
    Member member = memberReader.getMemberBy(request.getMemberId());
    Playlist playlist = playlistReader.getPlaylistBy(request.getPlaylistId());
    Optional<LikedPlaylist> likedPlaylist = likeReader.getLikedPlaylist(member, playlist);
    likedPlaylist.ifPresentOrElse(
        likeStore::delete,
        () -> likeStore.store(new LikedPlaylist(member, playlist))
    );

    return new LikedPlaylistInfo(request.getPlaylistId(), likedPlaylist.isEmpty());
  }

  @Override
  public LikedTrackInfo like(LikeTrackRequest request) {
    Member member = memberReader.getMemberBy(request.getMemberId());
    Track track = trackReader.getTrackBy(request.getTrackId());
    Optional<LikedTrack> likedTrack = likeReader.getLikedTrack(member, track);
    likedTrack.ifPresentOrElse(
        likeStore::delete,
        () -> likeStore.store(new LikedTrack(member, track))
    );

    return new LikedTrackInfo(request.getTrackId(), likedTrack.isEmpty());
  }
}

File: src/main/java/ordilov/lissn/like/domain/LikeServiceImpl.kt
package ordilov.lissn.like.domain

import ordilov.lissn.member.application.port.out.MemberReader
import ordilov.lissn.playlist.application.port.out.PlaylistReader
import ordilov.lissn.track.domain.TrackReader
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional
class LikeServiceImpl(
    val likeStore: LikeStore,
    val likeReader: LikeReader,
    val trackReader: TrackReader,
    val memberReader: MemberReader,
    val playlistReader: PlaylistReader
) : LikeService {

    override fun like(request: LikeTrackCommand): LikedTrackInfo {
        val member = memberReader.getMemberBy(request.memberId)
        val track = trackReader.getTrackBy(request.trackId)
        val likedTrack = likeReader.getLikedTrack(member, track)
        likedTrack.ifPresentOrElse(
            likeStore::delete
        ) { likeStore.store(LikedTrack(member, track)) }

        return LikedTrackInfo(request.trackId, likedTrack.isEmpty)
    }

    override fun like(request: LikePlaylistCommand): LikedPlaylistInfo {
        val member = memberReader.getMemberBy(request.memberId)
        val playlist = playlistReader.getPlaylistBy(request.playlistId)
        val likedPlaylist = likeReader.getLikedPlaylist(member, playlist)
        likedPlaylist.ifPresentOrElse(
            likeStore::delete
        ) { likeStore.store(LikedPlaylist(member, playlist)) }

        return LikedPlaylistInfo(request.playlistId, likedPlaylist.isEmpty)
    }

}